perun-core/src/main/java/cz/metacentrum/perun/core/impl/MembersManagerImpl.java

```diff
@@ -884,6 +906,47 @@ private String getSQLWhereForMembersPage(MembersPageQuery query, MapSqlParameter
 		return " AND " + Utils.prepareSqlWhereForUserMemberSearch(query.getSearchString(), namedParams, false);
 	}
 
+	private String getSQLBasedOnPolicy() {
+		return "LEFT OUTER JOIN (SELECT groups_members.member_id, authz.role_id, authz.vo_id" +
+			"		FROM groups" +
+			"			JOIN authz ON groups.id = authz.group_id" +
+			"			JOIN groups_members ON groups.id = groups_members.group_id" +
+			"		 WHERE authz.user_id = (:userId))" +
+			"		AS members_group ON members.id = members_group.member_id AND members.vo_id = members_group.vo_id";
+	}
+
+	private String getWhereConditionBasedOnThePolicy(PerunSession sess, MembersPageQuery query, String otherPolicy, Vo vo) throws PolicyNotExistsException {
+		PerunPolicy policy = AuthzResolverImpl.getPerunPolicy("filter-getMembersPage_policy");
+		if (otherPolicy != null && !otherPolicy.isEmpty()) {
+			policy = AuthzResolverImpl.getPerunPolicy(otherPolicy);
+		}
+
+		List<String> roles = new ArrayList<>();
+		for (Map<String,String> role : policy.getPerunRoles()) {
+			for (Map.Entry<String,String> entry : role.entrySet()) {
+				// Do nothing
+				if (entry.getValue() == null) continue;
+				if (entry.getValue().equals("Group")) {
+					int roleId = AuthzResolverBlImpl.getRoleIdByName(entry.getKey());
+					if (roleId == -1) {
+						log.error("Role {} not found in DB.", entry.getKey());
+						continue;
+					}
+					roles.add("members_group.role_id=" + roleId);
+				}
+			}
+		}
+
+		// Check if user is VO admin in vo
+		boolean ignoreGroupRelation = AuthzResolverBlImpl.isPerunAdmin(sess) || AuthzResolverBlImpl.isPerunObserver(sess) || isVoAdminOrObserver(sess, vo);
+		if (roles.isEmpty() || query.getGroupId() != null || ignoreGroupRelation) {
+			return " WHERE members.vo_id = (:voId)";
+		}
```

Comment:
We could do this before the for loop, no?

Reply:
Well, yes and no. The `roles.isEmpty()` is dependent on the outcome of the loop. There would be a kind of duplicity there. I mean, both of the solutions are not ideal.